File: apps/tauri/KawaCAD/src/features/parts/domain/partDefaults.ts
import { appStrings } from "@/localization";

export function partDefaultName(index: number): string {
  return `${appStrings.inspector.parts} ${index}`;
}

export function partLibraryPlacement(
  cursorPoint: { xMm: number; yMm: number } | undefined,
  existingOrigins: Array<{ xMm: number; yMm: number }>,
): { xMm: number; yMm: number } {
  if (cursorPoint) return cursorPoint;
  const maxY = existingOrigins.reduce((max, origin) => Math.max(max, origin.yMm), -Infinity);
  return {
    xMm: (existingOrigins.map((origin) => origin.xMm).reduce((max, value) => Math.max(max, value), -30) ?? -30) + 30,
    yMm: Number.isFinite(maxY) ? Math.max(maxY, 0) : 0,
  };
}

File: apps/macos/KawaCAD/Sources/KawaCADApp/Features/Parts/Domain/PartDefaults.swift
import Foundation

enum PartDefaults {
  static func name(number: Int) -> String {
    AppStrings.tr("part.default_name", number)
  }

  static func libraryPlacement(
    cursorPoint: ModelPoint?,
    existingOrigins: [ModelPoint]
  ) -> ModelPoint {
    if let cursorPoint {
      return cursorPoint
    }
    return ModelPoint(
      xMM: existingOrigins.reduce(-30) { max($0, $1.xMM) } + 30,
      yMM: existingOrigins.reduce(0) { max($0, $1.yMM) }
    )
  }
}
